File: Java/BasicAlgorithmPractice/src/StringRelatedExercises.java
import java.util.ArrayList;
import java.util.Arrays;

public class StringRelatedExercises {
    private String stringVar;

    private char charVar;

    public StringRelatedExercises(String stringVar, char charVar) {
        this.stringVar = stringVar;
        this.charVar = charVar;
    }

    public StringRelatedExercises(String stringVar) {
        this.stringVar = stringVar;
    }

    public boolean isAPalindrome() {
        String formattedString = this.stringVar.toLowerCase();
        formattedString = this.removeAllAlphanumericCharacters(formattedString);
        ArrayList<Character> listOfChars = this.getStringAsListArrayOfChars(formattedString);
        for (int i = 0; i < listOfChars.size() / 2; i++) {
            if (listOfChars.get(i) != listOfChars.get(listOfChars.size() - i - 1)) {
                return false;
            }
        }
        return true;
    }

    private ArrayList<Character> getStringAsListArrayOfChars(String strToConvert) {
        ArrayList<Character> listOfChars = new ArrayList<Character>();
        for (int i = 0; i < strToConvert.length(); i++) {
            listOfChars.add(strToConvert.charAt(i));
        }
        return listOfChars;
    }

    public String getStringVar() {
        return stringVar;
    }

    public char getCharVar() {
        return charVar;
    }

    public int getNumberOfOccurrences() {
        int numOfOccurrences = 0;
        ArrayList<Character> listOfChars = this.getStringAsListArrayOfChars(this.stringVar);
        for (Character character : listOfChars) {
            if (character == this.charVar) {
                numOfOccurrences++;
            }
        }

        return numOfOccurrences;
    }

    String removeAllSpaces(String string) {
        return string.replaceAll(" ","");
    }

    String removeAllAlphanumericCharacters(String string) {
        return string.replaceAll("[^a-zA-Z0-9]", "");
    }

    public String reverseString() {
        StringBuilder reversedString = new StringBuilder();
        ArrayList<Character> listOfChars = this.getStringAsListArrayOfChars(this.stringVar);
        for (int i = listOfChars.size() - 1; i > 0; i--) {
            reversedString.append(listOfChars.get(i));
        }
        reversedString.setLength(reversedString.length() - 1);
        return reversedString.toString();
    }

    public String titleCase() {
        String[] listOfWordsInString = this.getStringVar().split(" ");
        StringBuilder reformattedString = new StringBuilder();
        for (int i=0; i < listOfWordsInString.length; i++) {
            if (i > 0) {
                reformattedString.append(" ");
            }
            listOfWordsInString[i] =  listOfWordsInString[i].substring(0, 1).toUpperCase()
                    + listOfWordsInString[i].substring(1);
            reformattedString.append(listOfWordsInString[i]);
        }
        return reformattedString.toString();
    }
}

File: Kotlin/BasicAlgorithms/src/main/kotlin/strings/StringRelatedExercises.kt
package strings

import java.util.*

class StringRelatedExercises(
    private val stringVar: String,
    private val charVar: Char? = null
) {

    fun capitalizeAllWordsInString(): String {
        val stringArray =this.stringVar.split(" ")
        var stringOfCapitalizedLetters = ""
        for (i in 0..<stringArray.size) {
            stringOfCapitalizedLetters += capitalizeString(stringArray[i]) + " "
        }
        return stringOfCapitalizedLetters.substring(0, stringOfCapitalizedLetters.length - 1)
    }
    private fun capitalizeString(string: String): String {
        return string.replaceFirstChar {
            if (it.isLowerCase()) it.titlecase(Locale.getDefault())
            else
                it.toString()
        }
    }

    fun getCharVar():Char? {
        return charVar;
    }

    fun getNumberOfOccurrences(): Int {
        var numberOfOccurrences = 0;
        for (c in this.stringVar) {
            if(c == this.charVar) {
                numberOfOccurrences++;
            }
        }
        return numberOfOccurrences;
    }

    fun getStringVar(): String {
        return stringVar;
    }

    fun isAPalindrome(): Boolean {
        val formattedString = removeNonAlphanumericCharactersFromString(
            this.stringVar.lowercase()
        )
        for (i in 1..formattedString.length / 2) {
            if (formattedString[i - 1] != formattedString[formattedString.length - i]) {
                return false
            }
        }
        return true
    }

    private fun removeNonAlphanumericCharactersFromString(string: String): String {
        val nonAlphaNum = "[^a-zA-Z0-9]".toRegex()
        return string.replace(nonAlphaNum, "")
    }

    fun reverseString(): String {
        val reversedString = StringBuilder()
        for (i in this.stringVar.length - 1 downTo 0) {
            val char = this.stringVar[i]
            reversedString.append(char)
        }
        return reversedString.toString()
    }

}